sql-plugin/src/main/scala/com/nvidia/spark/rapids/limit.scala

```diff
@@ -58,61 +58,85 @@ trait GpuBaseLimitExec extends LimitExec with GpuExec with ShimUnaryExecNode {
     throw new IllegalStateException(s"Row-based execution should not occur for $this")
 
   override def doExecuteColumnar(): RDD[ColumnarBatch] = {
+    sliceRDD(child.executeColumnar(), limit, 0)
+  }
+
+  def sliceRDD(rdd: RDD[ColumnarBatch], limit: Int, offset: Int): RDD[ColumnarBatch] = {
+    val opTime = gpuLongMetric(OP_TIME)
     val numOutputRows = gpuLongMetric(NUM_OUTPUT_ROWS)
     val numOutputBatches = gpuLongMetric(NUM_OUTPUT_BATCHES)
-    val opTime = gpuLongMetric(OP_TIME)
-
-    val crdd = child.executeColumnar()
-    crdd.mapPartitions { cbIter =>
+    rdd.mapPartitions { iter =>
       new Iterator[ColumnarBatch] {
-        private var remainingLimit = limit
+        private var remainingLimit = limit - offset
+        private var remainingOffset = offset
 
-        override def hasNext: Boolean = remainingLimit > 0 && cbIter.hasNext
+        override def hasNext: Boolean = (limit == -1 || remainingLimit > 0) && iter.hasNext
 
         override def next(): ColumnarBatch = {
-          val batch = cbIter.next()
-          withResource(new NvtxWithMetrics("limit", NvtxColor.ORANGE, opTime)) { _ =>
-            val result = if (batch.numRows() > remainingLimit) {
-              sliceBatch(batch)
+          if (!iter.hasNext) {
+            throw new NoSuchElementException("Next on empty iterator")
+          }
+
+          var batch = iter.next()
+          val numCols = batch.numCols()
+
+          // In each partition, we need to skip `offset` rows
+          while (batch != null && remainingOffset >= batch.numRows()) {
+            remainingOffset -= batch.numRows()
+            batch.safeClose()
+            batch = if (iter.hasNext) { iter.next() } else { null }
+          }
+
+          // If the last batch is null, then we have offset >= numRows in this partition.
+          // In such case, we should return an empty batch
+          if (batch == null) {
+            return new ColumnarBatch(new ArrayBuffer[GpuColumnVector](numCols).toArray, 0)
+          }
+
+          // Here 0 <= remainingOffset < batch.numRow(), we need to get batch[remainingOffset:]
+          withResource(new NvtxWithMetrics("limit and offset", NvtxColor.ORANGE, opTime)) { _ =>
+            var result: ColumnarBatch = null
+            // limit < 0 (limit == -1) denotes there is no limitation, so when
+            // (remainingOffset == 0 && (remainingLimit >= batch.numRows() || limit < 0)) is true,
+            // we can take this batch completely
+            if (remainingOffset == 0 && (remainingLimit >= batch.numRows() || limit < 0)) {
+              result = batch
```

Comment:
Need to update the `remainingLimit ` in the `if` path.
```suggestion
              result = batch
              remainingLimit -= result.numRows()
```

Reply:
Or you can do this `remainingLimit -= result.numRows()` out of the `if else`.